cpp/include/raft/sparse/hierarchy/detail/mst.cuh

```diff
@@ -0,0 +1,188 @@
+
+  raft::sparse::convert::sorted_coo_to_csr(final_coo.rows(), final_coo.nnz,
+                                           indptr2.data(), m, d_alloc, stream);
+
+  value_idx max_offset = 0;
+  raft::update_host(&max_offset, indptr2.data() + (m - 1), 1, stream);
+  CUDA_CHECK(cudaStreamSynchronize(stream));
+
+  max_offset += (final_nnz - max_offset);
+
+  raft::update_device(indptr2.data() + m, &max_offset, 1, stream);
+
+  return raft::mst::mst<value_idx, value_idx, value_t>(
+    handle, indptr2.data(), final_coo.cols(), final_coo.vals(), m,
+    final_coo.nnz, color, stream, false, true);
+}
+
+/**
+ * Constructs an MST and sorts the resulting edges in ascending
+ * order by their weight.
+ *
+ * Hierarchical clustering heavily relies upon the ordering
+ * and vertices returned in the MST. If the result of the
+ * MST was actually a minimum-spanning forest, the CSR
+ * being passed into the MST is not connected. In such a
+ * case, this graph will be connected by performing a
+ * KNN across the components.
+ * @tparam value_idx
+ * @tparam value_t
+ * @param[in] handle raft handle
+ * @param[in] indptr CSR indptr of connectivities graph
+ * @param[in] indices CSR indices array of connectivities graph
+ * @param[in] pw_dists CSR weights array of connectivities graph
+ * @param[in] m number of rows in X / src vertices in connectivities graph
+ * @param[in] n number of columns in X
+ * @param[out] mst_src output src edges
+ * @param[out] mst_dst output dst edges
+ * @param[out] mst_weight output weights (distances)
+ */
+template <typename value_idx, typename value_t>
+void build_sorted_mst(const raft::handle_t &handle, const value_t *X,
+                      const value_idx *indptr, const value_idx *indices,
+                      const value_t *pw_dists, size_t m, size_t n,
+                      raft::mr::device::buffer<value_idx> &mst_src,
+                      raft::mr::device::buffer<value_idx> &mst_dst,
+                      raft::mr::device::buffer<value_t> &mst_weight,
+                      const size_t nnz) {
+  auto d_alloc = handle.get_device_allocator();
+  auto stream = handle.get_stream();
+
+  raft::mr::device::buffer<value_idx> color(d_alloc, stream, m);
+
+  auto mst_coo = raft::mst::mst<value_idx, value_idx, value_t>(
+    handle, indptr, indices, pw_dists, (value_idx)m, nnz, color.data(), stream,
+    false);
+
+  if (linkage::get_n_components(color.data(), m, stream) > 1) {
+    mst_coo = connect_knn_graph<value_idx, value_t>(handle, X, mst_coo, m, n,
+                                                    color.data());
+
+    printf("Edges: %d\n", mst_coo.n_edges);
+
+    RAFT_EXPECTS(
+      mst_coo.n_edges == m - 1,
+      "MST was not able to connect knn graph in a single iteration.");
+  }
+
+  sort_coo_by_data(mst_coo.src.data(), mst_coo.dst.data(),
+                   mst_coo.weights.data(), mst_coo.n_edges, stream);
+
+  // TODO: be nice if we could pass these directly into the MST
```

Comment:
This is where you want to be able to pass a dense input, right?

Reply:
No, the unfortunate side effect of the pointers being constructed on the heap by the function being called is that I have no direct control over the memory being used. What if I wanted to hand you a specific chunk of memory to use? This is one of the reasons I'm in favor of handing the function an object that hasn't yet been initialized and allowing the function to call an "initialize" (specifically in the case of sparse data, where the nnz might not be known ahead of time). It allows me to maintain ownership over the object being constructed from the start, and temporarily grants the ability for the called function to mutate that object. My problem with this inversion of ownership is that now it leaves me having to return an object from this function that was created in this function and that propagates to all functions that need to use the resulting data. I'd much rather create an object in the caller and propagate that same object through the call stack  (as we do with the `raft::sparse:COO` everywhere.